Docs style guide (#1030)
SAP/cloud-sdk: style-guide.md

> # About This Guide

Reviewer: [pp] I think it would be nice to be consistent with the naming of our files. In the JS repo, we call it `STYLEGUIDE.md`.

Reviewer: [nth] I think it would be helpful to have a table of contents same as here: (https://github.com/SAP/cloud-sdk-js/blob/main/STYLEGUIDE.md#table-of-contents)

> # About This Guide
> 
> This style guide covers some general text-formatting conventions that should be applied when writing documentation for the SAP Cloud SDK.
> 
> ## Formatting

Reviewer: [req] The highest level title should be the title of the document. Currently "Formatting" us a section under "About This Guide".
Suggestion: remove the first level and drag all sections one level up.
```suggestion
# Formatting

This style guide covers some general text-formatting conventions that should be applied when writing documentation for the SAP Cloud SDK.
```

### Code Format for Inline Code

> Use code font to mark up code, such as class names, method names, HTTP status codes/verbs/content-types, env variables, console output, etc.

Reviewer: [req] I think this should be a list:
```suggestion
Use code font to mark up code, such as:
- class names
- method names
- HTTP status codes/verbs/content-types
- env variables
- console output
- etc.
```

✅ instance of the `BusinessPartner` entity class

> For a complete list, see [here](https://developers.google.com/style/code-in-text#some-specific-items-to-put-in-code-font)

Reviewer: [req] I think this section would benefit from a subsection what not to put into code font.

> 

Reviewer: [req] I think we should have a subsection here for the special case of methods (and functions) => https://developers.google.com/style/code-in-text#method-names
[nth] Maybe it would also be nice to have non-formatting related section below on when to speak of methods and when to call something a function. I have the impression that this is still misunderstood a lot.

### Placeholder Text Syntax

Use all caps with underscore delimiters.
[Explain your placeholders](https://developers.google.com/style/placeholders#explain-placeholders).

✅ PLACEHOLDER_NAME

❌ PLACEHOLDER-Name

❌ Placeholder-Name

<!-- vale off -->

❌ placeholderName

> <!-- vale on -->

Reviewer: I think we should also include the rule that possessive pronouns should not be included.

> ### User-Replaced Placeholders in Inline Text

Reviewer: [req] I find it a bit confusing that this section and the next section are on the same level as the general place holder section. Suggestion:
Have 1 section on "Placeholders" with some general guidance and below that add 3 sections on:
- Placeholders in Code
- Placeholders in Text
- Placeholder Text (or Placeholder Syntax, but I slightly lean towards Text)

To me that would be the natural order of how I think about that. I first have a code example. Then I have the text that I use to explain the example. From the examples in the guide I probably already have an idea of how the placeholder text should look like, but if I want more details, there is a detailed section on this.

When placeholders occur in a sentence use code format in italics.

> ✅ where _`YOUR_APPLICATION_NAME`_ is replaced by the name you specified.

Reviewer 2: general: I thought we should use active language. If so would be nice if the examples feature that.

Author: @FrankEssenberger I used the examples from our text for the formatting so it was not super generic examples but provided more clarity. Since this guide was supposed to be strictly for formatting (scope of the ticket) and not about language and writing style, I decided not to go there.

I can however add a small section still with some examples.

Reviewer: [pp] But I agree, we could easily change this example to active voice without making this part of the guide.
```suggestion
✅ Replace _`YOUR_APPLICATION_NAME`_ with your application name.
```

Reviewer: [req] However, according to the google guide (and I like this rule) we should not use possessive pronouns, so maybe this example should not use a possessive pronoun either ;)

❌ where `<YOUR-APPLICATION-NAME>` is replaced by the name you specified.

Sometimes, italics can be skipped (URL, inline code sample).

✅ as in a JWT issued from the XSUAA, e.g., `https://YOUR_SUBDOMAIN.localhost:8080/uaa/oauth/token`.

❌ as in a JWT issued from the XSUAA, e.g., `https://yourSubdomain.localhost:8080/uaa/oauth/token`.

### User-Replaced Placeholders in Code Block

Use the normal placeholder syntax.

✅

```YAML
applications:
  - name: YOUR_APPLICATION_NAME
    path: .
```

❌

```YAML
applications:
  - name: <YOUR-APPLICATION-NAME>
    path: .
```

> ### Software Product and their Command Line Utility

Reviewer: [nth]
```suggestion
### Software Products and Command Line Utilities
```

Reviewer: [pp] I am not sure that these two things need to be put together. Command line utilities could just be mentioned in the "Code Format for Inline Code" or be a separate section.

Use code font for the command and ordinary font for the name of the project or product.

✅ You will need the Cloud Foundry CLI (`cf`)

❌ You will need the **Cloud Foundry CLI** (cf)

### URL and Domains in Text

Use code font.

✅ against a destination with URL `https://my.s4-system.com`

### Filenames and FilePaths in Text

Use code font.
Include the word file after the filename.

> ✅ Add your own `package.json` file.

Reviewer: [nth] Maybe add an example for a file path as well?

### File Type in Text

✅ metadata specification in the EDMX format

❌ metadata specification in the `EDMX` format

> Try to use the formal file type name, not the filename extension.

Reviewer: [req] Can you give an example for this? e.g. XML file vs. .xml file?

### Error Messages in Text

Sentence-style capitalization.

> Enclose error messages in quotation marks when referring to them in text.

Reviewer: [q] I think there can be quite complex cases where the error already includes both " and '. How should this be handled?

Author: I assume the same. Also I didn't notice any issues in markdown syntax if you add text that includes. When the error message is too complex/large, we can use the  fenced block syntax (which I have seen being used)

✅ If you see the error "Module not found: Error: Can't resolve 'https' in"

❌ If you see the error `Module not found: Error: Can't resolve 'https' in`

### Example Domains

> For generic domain names in examples, use `example.com`, `example.org`, or `example.net`.

Reviewer: [q] Should we restrict this to only `example.com`? I doubt that all three will be needed.
[q] Also, what do I do if I need to use two different domains. I think `example.com` are `example.org` too similar. I guess then something like `other.domain.com` would be OK (or would it not?). Should we document this or rather not?
[q] Strictly speaking this is not about formatting, but naming. Should we move this down to the other non-formatting related aspects?

Author: [1] 👍🏽 
[2] I honestly dont have a good suggestion here. We can leave this point open and document it when we do have such a scenario and have made a decision what to use.  
[3] I agree, but I would keep it here because It also doesn't fit under Language and Style and Punctuation either. Later if we have a general section, we can move it there.

### Graphical User Interface Elements

> Bold all graphical user interface element names, including, menus, button, dialog box, list items, and window.

Reviewer: [req] I think this should be a list for better readability.

Do not use italics or quotes.

✅ On the **Settings** page, click **Show API Key**

❌ On the settings page, click _Show API Key_.

✅ In the dropdown, choose **BasicAuthentication** as the authentication type

❌ In the dropdown, choose `BasicAuthentication` as the authentication type

> Don't include the type of UI element, such as button or checkbox, unless it adds needed clarity.

Reviewer: [pp] I think we should either drag this text up to the top of this section or introduce subsections for better overview. I don't know what the best option here is...

> Use sentence-style capitalization unless you need to match the UI.

Reviewer: [q] When do I not need to match the UI? I really don't understand when this would be the case.

Author: When either the labels themselves are inconsistent or all caps, in that case, change all to sentence style. I improved the text a bit to avoid confusion.

If a label ends with a colon/an ellipsis, don't include that end punctuation.

> ✅ To start your application, click **Launch**.
> 
> ❌ You can start your application via the "launch.." button.
> 
> ❌ You can start your application via the "LAUNCH" button.

Reviewer: [req] I think these examples are not enough to understand the requirements well. The main problem is, that we do not see any UI with a button, therefore adding info on the button label would be helpful.
```suggestion
Button label: launch...

✅ To start your application, click **launch**.

✅ **Launch** starts your application.

❌ To start your application, click **launch...**.

❌ To start your application, click **LAUNCH**.

❌ You can start your application via the "launch.." button.
```

### User Entered Text

Use code format for text that should be entered by user

✅ In the **Name** field, enter `my-destination`

❌ In the **Name** field, enter my-destination

> ## Avoid Using We

Reviewer: [q] I think from here on this is not formatting anymore, maybe this should have another separate heading like "Style"?

> ✅ More than 100 deprecated methods have been removed.

Reviewer 2: [q] how do you avoid `we` and also use active language?

Author: In some cases, there can be exceptions as explained [here](https://developers.google.com/style/voice#exceptions) (To de-emphasize a subject or actor.).

Reviewer 2: ok, this is good to know. For me as a non native speaker it is hard to judge when it is ok to use passiv form, but this is the problem with normal language with their non strict rules

❌ We removed a bit more than 100 deprecated methods.

✅ The SAP Cloud SDK does not guarantee any API compatibility for future updates

❌ We do not guarantee any API compatibility for future updates

## Version

Unless a text relates to a specific version, do not use formats x.y or x.y.z

✅ This mapping applies to the move to open source version 1.

<!-- vale off -->

❌ This mapping applies to the move to open source version 1.X.Y..

<!-- vale on -->

Avoid words like "currently" and "new".

✅ In version 2, we allow to specify the origin of the header

> ❌ In the new version we allow to specify the origin of the header

Reviewer: [req] Remove "we" from these samples ;)
